infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/Orchestrator.java

```diff
@@ -0,0 +1,153 @@
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.corfudb.format.Types.OrchestratorRequestType.QUERY;
+
+/**
+ * The orchestrator is a stateless service that runs on all management servers and its purpose
+ * is to execute workflows. A workflow defines multiple smaller actions that must run in order
+ * that is specified by Workflow.getActions() to achieve a bigger goal. For example, growing the
+ * cluster. Initiated through RPC, the orchestrator will create a workflow instance and attempt
+ * to execute all its actions.
+ *
+ * <p>
+ * Created by Maithem on 10/25/17.
+ */
+
+@Slf4j
+public class Orchestrator {
+
+    final Callable<CorfuRuntime> getRuntime;
+    final Set<UUID> activeWorkflows = ConcurrentHashMap.newKeySet();
+
+    public Orchestrator(@Nonnull Callable<CorfuRuntime> runtime) {
+        this.getRuntime = runtime;
+    }
+
+    public void handle(@Nonnull CorfuPayloadMsg<OrchestratorRequest> msg,
+                       @Nonnull ChannelHandlerContext ctx,
+                       @Nonnull IServerRouter r) {
+
+        OrchestratorRequest orchReq = msg.getPayload();
+
+        if (orchReq.getRequest().getType() == QUERY) {
+            handleQuery(msg, ctx, r);
+        } else {
+            dispatch(msg, ctx, r);
+        }
+    }
+
+    void handleQuery(CorfuPayloadMsg<OrchestratorRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
+        QueryRequest req = (QueryRequest) msg.getPayload().getRequest();
+
+        Response resp;
+        if (activeWorkflows.contains(req.getId())) {
+            resp = new QueryResponse(true);
+        } else {
+            resp = new QueryResponse(false);
+        }
+
+        r.sendResponse(ctx, msg, CorfuMsgType.ORCHESTRATOR_RESPONSE
+                .payloadMsg(new OrchestratorResponse(resp )));
+    }
+
+    void dispatch(CorfuPayloadMsg<OrchestratorRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
+        CompletableFuture.runAsync(() -> {
+            OrchestratorRequest req = msg.getPayload();
+            Workflow workflow = getWorkflow(req);
+            OrchestratorResponse resp = new OrchestratorResponse(new AddNodeResponse(workflow.getId()));
+            r.sendResponse(ctx, msg, CorfuMsgType.ORCHESTRATOR_RESPONSE
+                    .payloadMsg(resp));
+            System.out.println("Sent add node response");
+            run(workflow);
+        });
+    }
+
+    /**
+     * Create a workflow instance from an orchestrator request
+     * @param req Orchestrator request
+     * @return Workflow instance
+     */
+    @Nonnull
+    private Workflow getWorkflow(@Nonnull OrchestratorRequest req) {
+        Request payload = req.getRequest();
+        if (payload.getType().equals(Types.OrchestratorRequestType.ADD_NODE)) {
```

Comment:
You might need better deduping here. If there is already an add_node for the same node going on should you create another workflow for that ?

Reply:
Ok. Will return the same workflow UUID if one is running for the same endpoint

Reply:
done